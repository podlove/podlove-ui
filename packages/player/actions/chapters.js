import { createAction } from './action-creator'

import {
  SET_CHAPTERS_LIST,
  NEXT_CHAPTER,
  PREVIOUS_CHAPTER,
  SET_CHAPTER,
  UPDATE_CHAPTER
} from './types'

export const setChapters = createAction(SET_CHAPTERS_LIST)
export const nextChapter = createAction(NEXT_CHAPTER)
export const previousChapter = createAction(PREVIOUS_CHAPTER)
export const setChapter = createAction(SET_CHAPTER)
export const updateChapter = createAction(UPDATE_CHAPTER)
