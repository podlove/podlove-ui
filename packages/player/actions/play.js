import { createAction } from './action-creator'

import {
  REQUEST_PLAY,
  REQUEST_PAUSE,
  REQUEST_RESTART,
  BACKEND_PLAY,
  BACKEND_PAUSE,
  BACKEND_END,
  BACKEND_LOADING_START,
  BACKEND_LOADING_END,
  BACKEND_BUFFER,
  BACKEND_ERROR
} from './types'

export const requestPlay = createAction(REQUEST_PLAY)
export const backendPlay = createAction(BACKEND_PLAY)

export const requestPause = createAction(REQUEST_PAUSE)
export const backendPause = createAction(BACKEND_PAUSE)

export const requestRestart = createAction(REQUEST_RESTART)

export const backendLoadingStart = createAction(BACKEND_LOADING_START)
export const backendLoadingEnd = createAction(BACKEND_LOADING_END)
export const backendBuffer = createAction(BACKEND_BUFFER)
export const backendEnd = createAction(BACKEND_END)

export const backendError = createAction(BACKEND_ERROR)
