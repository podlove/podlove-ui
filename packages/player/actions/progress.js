import { ENABLE_GHOST_MODE, DISABLE_GHOST_MODE } from './types'
import { createAction } from './action-creator'

export const enableGhost = createAction(ENABLE_GHOST_MODE)
export const disableGhost = createAction(DISABLE_GHOST_MODE)
