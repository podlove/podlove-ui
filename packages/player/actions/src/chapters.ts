import { createAction } from 'redux-actions';
import type { PodloveWebPlayerChapter } from '@podlove/types';

import {
  SET_CHAPTERS_LIST,
  NEXT_CHAPTER,
  PREVIOUS_CHAPTER,
  SET_CHAPTER,
  UPDATE_CHAPTER
} from './types';

export type setChaptersPayload = PodloveWebPlayerChapter[];
export type setChapterPayload = number;
export type updateChapterPayload = number;

export const setChapters = createAction<setChaptersPayload>(SET_CHAPTERS_LIST);
export const nextChapter = createAction<void>(NEXT_CHAPTER);
export const previousChapter = createAction<void>(PREVIOUS_CHAPTER);
export const setChapter = createAction<setChapterPayload>(SET_CHAPTER);
export const updateChapter = createAction<updateChapterPayload>(UPDATE_CHAPTER);
