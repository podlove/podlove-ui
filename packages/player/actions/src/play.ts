import { createAction } from 'redux-actions';
import type { PodloveWebPlayerAudioError, PodloveWebPlayerAudioProps } from '@podlove/types';

import {
  REQUEST_PLAY,
  REQUEST_PAUSE,
  REQUEST_RESTART,
  BACKEND_PLAY,
  BACKEND_PAUSE,
  BACKEND_END,
  BACKEND_LOADING_START,
  BACKEND_LOADING_END,
  BACKEND_BUFFER,
  BACKEND_ERROR,
  REQUEST_LOAD,
  REQUEST_STOP
} from './types';

export type requestPlayPayload = void;
export type backendPlayPayload = void;
export type requestPausePayload = void;
export type requestLoadPayload = void;
export type backendPausePayload = void;
export type requestStopPayload = void;
export type requestRestartPayload = void;
export type backendLoadingStartPayload = void;
export type backendLoadingEndPayload = PodloveWebPlayerAudioProps;
export type backendBufferPayload = [number, number][];
export type backendEndPayload = void;
export type backendErrorPayload = PodloveWebPlayerAudioError;

export const requestPlay = createAction<requestPlayPayload>(REQUEST_PLAY);
export const backendPlay = createAction<backendPlayPayload>(BACKEND_PLAY);

export const requestPause = createAction<requestPausePayload>(REQUEST_PAUSE);
export const requestStop = createAction<requestStopPayload>(REQUEST_STOP);
export const requestLoad = createAction<requestLoadPayload>(REQUEST_LOAD);
export const backendPause = createAction<backendPausePayload>(BACKEND_PAUSE);

export const requestRestart = createAction<requestRestartPayload>(REQUEST_RESTART);

export const backendLoadingStart = createAction<backendLoadingStartPayload>(BACKEND_LOADING_START);
export const backendLoadingEnd = createAction<backendLoadingEndPayload>(BACKEND_LOADING_END);
export const backendBuffer = createAction<backendBufferPayload>(BACKEND_BUFFER);
export const backendEnd = createAction<backendEndPayload>(BACKEND_END);

export const backendError = createAction<backendErrorPayload>(BACKEND_ERROR);
