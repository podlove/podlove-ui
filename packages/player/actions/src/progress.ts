import { ENABLE_GHOST_MODE, DISABLE_GHOST_MODE } from './types.js';
import { createAction } from 'redux-actions';

export type enableGhostPayload = void;
export type disableGhosttPayload = void;

export const enableGhost = createAction<enableGhostPayload>(ENABLE_GHOST_MODE);
export const disableGhost = createAction<disableGhosttPayload>(DISABLE_GHOST_MODE);
