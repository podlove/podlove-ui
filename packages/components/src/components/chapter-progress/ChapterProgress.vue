<template>
  <div
    ref="progressContainer"
    class="relative flex items-center px-2 py-0"
    aria-hidden="true"
    @mouseout="progressOut"
    @mousemove.alt="progressMove"
    @click.exact="progressClick"
    @click.alt="contextProgressClick"
  >
    <span class="title pointer-events-none" aria-hidden="true">
      {{ chapter.title }}
    </span>
    <span v-if="hasLink" class="link flex">
      <icon class="icon flex-shrink-0" type="link" />
      <a
        class="font-medium whitespace-no-wrap"
        :href="chapter.href"
        target="_blank"
        @mouseover="linkOver"
        @mouseleave="linkLeave"
      >
        {{ chapter.linkTitle }}
      </a>
    </span>
    <timer
      class="timer block text-right pointer-events-none pr-2"
      :time="remainingTime"
      :remaining="(chapter.active || !!ghostActive) && playtime > 0"
    />
    <span
      class="chapter-progress-bar absolute left-0 bottom-0 pointer-events-none"
      :style="progressStyle"
      aria-hidden="true"
    />
    <span
      class="chapter-progress-bar absolute left-0 bottom-0 pointer-events-none"
      :style="progressGhostStyle"
      aria-hidden="true"
    />
  </div>
</template>

<script>
import { disableGhost, enableGhost } from '@podlove/player-actions/progress'
import { setChapter } from '@podlove/player-actions/chapters'
import { simulatePlaytime, requestPlaytime } from '@podlove/player-actions/timepiece'
import { requestPlay } from '@podlove/player-actions/play'

import { background } from '../../defaults'
import { fade } from 'farbraum'
import Icon from '../icons/Icon'
import Timer from '../timer/Timer'

export default {
  components: {
    Icon,
    Timer
  },
  props: {
    chapter: {
      type: Object,
      default: () => ({
        start: 0,
        end: 0,
        title: '',
        href: null,
        linkTitle: null,
        active: false
      })
    },
    showLink: {
      type: Boolean,
      default: false
    },
    playtime: {
      type: Number,
      default: 0
    },
    ghost: {
      type: Number,
      default: 0
    },
    progressColor: {
      type: String,
      default: background
    }
  },

  computed: {
    progressStyle() {
      if (!this.chapter.active || this.playtime > this.chapter.end) {
        return {}
      }

      return {
        width: this.progress(this.playtime),
        'background-color': this.progressColor
      }
    },

    ghostActive() {
      return this.ghost && this.ghost > this.chapter.start && this.ghost < this.chapter.end
    },

    progressGhostStyle() {
      if (!this.ghostActive) {
        return {}
      }

      return {
        width: this.progress(this.ghost),
        'background-color': fade(this.progressColor, 0.7)
      }
    },

    remainingTime() {
      if (this.chapter.active) {
        return this.chapter.end - this.playtime
      }

      if (this.ghostActive) {
        return this.chapter.end - this.ghost
      }

      return this.chapter.end - this.chapter.start
    },

    hasLink() {
      return this.chapter.href && this.showLink
    }
  },

  methods: {
    hoverPlaytime(event) {
      const clientRect = this.$refs.progressContainer.getBoundingClientRect()
      return (
        this.chapter.start +
        ((this.chapter.end - this.chapter.start) * (event.clientX - clientRect.left)) /
          clientRect.width
      )
    },

    progress(time) {
      return `${((time - this.chapter.start) * 100) / (this.chapter.end - this.chapter.start)}%`
    },

    progressClick() {
      this.$emit('chapter', setChapter(this.chapter.index - 1))
      this.$emit('play', requestPlay())
      return false
    },

    contextProgressClick(event) {
      this.$emit('playtime', requestPlaytime(this.hoverPlaytime(event)))
      this.$emit('play', requestPlay())
      event.preventDefault()
      return false
    },

    progressOut() {
      this.$emit('ghost', disableGhost())
    },

    progressMove(event) {
      this.$emit('ghost', enableGhost())

      this.$emit('simulate', simulatePlaytime(this.hoverPlaytime(event)))
    },

    linkOver() {
      this.$emit('hover', true)
    },

    linkLeave() {
      this.$emit('hover', false)
    }
  }
}
</script>

<style lang="scss" scoped>
.title {
  width: calc(100% - 4.4em);
}

.link {
  max-width: calc(40%);
}

.icon {
  margin-top: -2px;
}

.timer {
  min-width: 4.4em;
}

.chapter-progress-bar {
  height: 3px;
}
</style>
