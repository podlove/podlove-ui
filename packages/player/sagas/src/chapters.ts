import { propOr, prop } from 'ramda';
import { takeEvery, select, put } from 'redux-saga/effects';
import type { Action } from 'redux-actions';

import { ready, type readyPayload } from '@podlove/player-actions/lifecycle'
import * as config from '@podlove/player-config';
import * as chapter from '@podlove/player-actions/chapters';
import { backendPlaytime, requestPlaytime, type requestPlaytimePayload } from '@podlove/player-actions/timepiece';
import { toPlayerTime } from '@podlove/utils/time';
import { hostname } from '@podlove/utils/url';
import type { PodloveWebPlayerChapter } from '@podlove/types'
import { disableGhost } from '@podlove/player-actions/progress'

export const chaptersSaga = ({
  selectDuration,
  selectPlaytime,
  selectCurrentChapter,
  selectChapterList
}: {
  selectDuration: (input: any) => number;
  selectPlaytime: (input: any) => number;
  selectCurrentChapter: (input: any) => PodloveWebPlayerChapter;
  selectChapterList: (input: any) => PodloveWebPlayerChapter[];
}) =>
  function* saga() {
    yield takeEvery(ready.toString(), initChapters, {
      selectDuration
    });
    yield takeEvery(requestPlaytime.toString(), chapterUpdate);
    yield takeEvery(backendPlaytime.toString(), chapterUpdate);
    yield takeEvery(chapter.setChapter.toString(), setCurrentChapter, {
      selectCurrentChapter
    });
    yield takeEvery(disableGhost.toString(), resetChapter, {
      selectPlaytime
    });
    yield takeEvery(chapter.previousChapter.toString(), previousChapter, {
      selectPlaytime,
      selectCurrentChapter
    });
    yield takeEvery(chapter.nextChapter.toString(), nextChapter, {
      selectCurrentChapter,
      selectChapterList
    });
  };

export function* chapterUpdate({ payload }: Action<requestPlaytimePayload>) {
  yield put(chapter.updateChapter(payload));
}

export function* setCurrentChapter({ selectCurrentChapter }: { selectCurrentChapter: (input: any) => PodloveWebPlayerChapter}): any {
  const current: PodloveWebPlayerChapter = yield select(selectCurrentChapter);

  if (!current.start) {
    return;
  }

  yield put(requestPlaytime(current.start));
}

export function* resetChapter({ selectPlaytime }: { selectPlaytime: (input: any) => number }): any {
  const playtime: number = yield select(selectPlaytime);
  yield put(chapter.updateChapter(playtime));
}

export function* previousChapter({ selectPlaytime, selectCurrentChapter }: { selectPlaytime: (input: any) => number, selectCurrentChapter: (input: any) => PodloveWebPlayerChapter }): any {
  const playtime: number = yield select(selectPlaytime);
  const { start, index }: PodloveWebPlayerChapter = yield select(selectCurrentChapter);

  if (!start) {
    return
  }

  if (playtime - start <= 1000) {
    yield put(chapter.setChapter(index - 2 < 0 ? 0 : index - 2));
  } else {
    yield put(chapter.setChapter(index - 1 < 0 ? 0 : index - 1));
  }
}

export function* nextChapter({ selectChapterList, selectCurrentChapter }: { selectChapterList: (input: any) => PodloveWebPlayerChapter[], selectCurrentChapter: (input: any) => PodloveWebPlayerChapter }): any {
  const chapters: PodloveWebPlayerChapter[] = yield select(selectChapterList);
  const { index }: PodloveWebPlayerChapter = yield select(selectCurrentChapter);

  yield put(chapter.setChapter(index > chapters.length - 1 ? chapters.length - 1 : index));
}

export function* initChapters({ selectDuration }: { selectDuration: (input: any) => number }, { payload }: Action<readyPayload>) {
  const chapters = config.chapters(payload);
  const duration: number = yield select(selectDuration);

  const state = chapters.reduce((result: PodloveWebPlayerChapter[], chapter: PodloveWebPlayerChapter, index: number, chapters: PodloveWebPlayerChapter[]) => {
    const end: PodloveWebPlayerChapter = propOr({ start: duration } as PodloveWebPlayerChapter, (index + 1).toString(), chapters);
    const href = (propOr('', 'href', chapter) as string).trim();

    return [
      ...result,
      {
        index: index + 1,
        start: toPlayerTime(chapter?.start || 0),
        end: toPlayerTime(end?.start || 0),
        title: prop('title', chapter),
        image: prop('image', chapter),
        href,
        linkTitle: href ? hostname(href) : null
      }
    ];
  }, []);

  yield put(chapter.setChapters(state));
}
