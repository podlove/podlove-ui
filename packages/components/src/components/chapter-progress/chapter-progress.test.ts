import { describe, test, expect } from 'vitest';
import { mount } from '@vue/test-utils'
import { setChapter } from '@podlove/player-actions/chapters'
import { disableGhost } from '@podlove/player-actions/progress'
import { requestPlay } from '@podlove/player-actions/play'

import ChapterProgress from './ChapterProgress.vue'

describe('ChapterProgress', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(ChapterProgress)

      expect(wrapper.vm).toBeTruthy()
    })
  })

  describe('props', () => {
    test('should render without props', async () => {
      const wrapper = await mount(ChapterProgress)
      expect(wrapper.element).toMatchSnapshot()
    })

    const props = [
      {
        prop: 'chapter',
        value: {
          start: 0,
          end: 50,
          title: 'chapter title',
          href: 'http://foo.bar',
          linkTitle: 'link title',
          active: true
        }
      },
      {
        prop: 'showLink',
        value: true
      },
      {
        prop: 'playtime',
        value: 25
      },
      {
        prop: 'progressColor',
        value: '#fff'
      }
    ]

    props.forEach(({ prop, value }) => {
      test(`should render '${prop}'`, async () => {
        const wrapper = await mount(ChapterProgress, {
          props: {
            [prop]: value
          }
        })

        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('computed', () => {
    test('should render the remaining time accordingly', async () => {
      const chapterActive = await mount(ChapterProgress, {
        props: {
          chapter: {
            start: 0,
            end: 50,
            title: 'chapter title',
            href: 'http://foo.bar',
            linkTitle: 'link title',
            active: true
          },
          playtime: 20
        }
      })

      expect(chapterActive.componentVM.remainingTime).toEqual(30)

      const ghostActive = await mount(ChapterProgress, {
        props: {
          chapter: {
            start: 0,
            end: 50,
            title: 'chapter title',
            href: 'http://foo.bar',
            linkTitle: 'link title',
            active: false
          },
          ghost: 30,
          ghostActive: true
        }
      })

      expect(ghostActive.componentVM.remainingTime).toEqual(20)

      const defaultActive = await mount(ChapterProgress, {
        props: {
          chapter: {
            start: 0,
            end: 50,
            title: 'chapter title',
            href: 'http://foo.bar',
            linkTitle: 'link title',
            active: false
          }
        }
      })

      expect(defaultActive.componentVM.remainingTime).toEqual(50)
    })
  })

  describe('events', () => {
    test('should emit SET_CHAPTER on progress click', async () => {
      const wrapper = await mount(ChapterProgress, {
        props: {
          chapter: {
            start: 0,
            end: 50,
            title: 'chapter title',
            href: 'http://foo.bar',
            linkTitle: 'link title',
            active: true,
            index: 2
          }
        }
      })

      wrapper.trigger('click')
      expect(wrapper.emitted().chapter).toEqual([[setChapter(1)]])
    })

    test('should emit REQUEST_PLAY on progress click', async () => {
      const wrapper = await mount(ChapterProgress, {
        props: {
          chapter: {
            start: 0,
            end: 50,
            title: 'chapter title',
            href: 'http://foo.bar',
            linkTitle: 'link title',
            active: true,
            index: 2
          }
        }
      })

      wrapper.trigger('click')
      expect(wrapper.emitted().play).toEqual([[requestPlay()]])
    })

    test('should emit DISABLE_GHOST on mouse out', async () => {
      const wrapper = await mount(ChapterProgress, {
        props: {
          chapter: {
            start: 0,
            end: 50,
            title: 'chapter title',
            href: 'http://foo.bar',
            linkTitle: 'link title',
            active: true,
            index: 2
          }
        }
      })

      wrapper.trigger('mouseout')
      expect(wrapper.emitted().ghost).toEqual([[disableGhost()]])
    })
  })
})
