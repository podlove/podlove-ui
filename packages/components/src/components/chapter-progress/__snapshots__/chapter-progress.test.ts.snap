// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChapterProgress > props > should render 'chapter' 1`] = `
<div
  aria-hidden="true"
  class="podlove-component-chapter-progress relative flex items-center px-2 py-0"
  data-v-1cf33334=""
>
  <span
    aria-hidden="true"
    class="pointer-events-none w-[calc(100%-4.4em)]"
    data-v-1cf33334=""
  >
    chapter title
  </span>
  <!--v-if-->
  <time
    class="podlove-timer block text-right pointer-events-none pr-2 min-w-[4.4em]"
    data-v-1cf33334=""
    datetime="00:00"
  >
    00:00
  </time>
  <span
    aria-hidden="true"
    class="absolute left-0 bottom-0 pointer-events-none h-[3px] podlove-component-chapter-progress-bar"
    data-v-1cf33334=""
  />
</div>
`;

exports[`ChapterProgress > props > should render 'playtime' 1`] = `
<div
  aria-hidden="true"
  class="podlove-component-chapter-progress relative flex items-center px-2 py-0"
  data-v-1cf33334=""
>
  <span
    aria-hidden="true"
    class="pointer-events-none w-[calc(100%-4.4em)]"
    data-v-1cf33334=""
  />
  <!--v-if-->
  <time
    class="podlove-timer block text-right pointer-events-none pr-2 min-w-[4.4em]"
    data-v-1cf33334=""
    datetime="00:00"
  >
    00:00
  </time>
  <span
    aria-hidden="true"
    class="absolute left-0 bottom-0 pointer-events-none h-[3px]"
    data-v-1cf33334=""
  />
</div>
`;

exports[`ChapterProgress > props > should render 'progressColor' 1`] = `
<div
  aria-hidden="true"
  class="podlove-component-chapter-progress relative flex items-center px-2 py-0"
  data-v-1cf33334=""
  progresscolor="#fff"
>
  <span
    aria-hidden="true"
    class="pointer-events-none w-[calc(100%-4.4em)]"
    data-v-1cf33334=""
  />
  <!--v-if-->
  <time
    class="podlove-timer block text-right pointer-events-none pr-2 min-w-[4.4em]"
    data-v-1cf33334=""
    datetime="00:00"
  >
    00:00
  </time>
  <span
    aria-hidden="true"
    class="absolute left-0 bottom-0 pointer-events-none h-[3px]"
    data-v-1cf33334=""
  />
</div>
`;

exports[`ChapterProgress > props > should render 'showLink' 1`] = `
<div
  aria-hidden="true"
  class="podlove-component-chapter-progress relative flex items-center px-2 py-0"
  data-v-1cf33334=""
>
  <span
    aria-hidden="true"
    class="pointer-events-none w-[calc(100%-4.4em)]"
    data-v-1cf33334=""
  />
  <!--v-if-->
  <time
    class="podlove-timer block text-right pointer-events-none pr-2 min-w-[4.4em]"
    data-v-1cf33334=""
    datetime="00:00"
  >
    00:00
  </time>
  <span
    aria-hidden="true"
    class="absolute left-0 bottom-0 pointer-events-none h-[3px]"
    data-v-1cf33334=""
  />
</div>
`;

exports[`ChapterProgress > props > should render without props 1`] = `
<div
  aria-hidden="true"
  class="podlove-component-chapter-progress relative flex items-center px-2 py-0"
  data-v-1cf33334=""
>
  <span
    aria-hidden="true"
    class="pointer-events-none w-[calc(100%-4.4em)]"
    data-v-1cf33334=""
  />
  <!--v-if-->
  <time
    class="podlove-timer block text-right pointer-events-none pr-2 min-w-[4.4em]"
    data-v-1cf33334=""
    datetime="00:00"
  >
    00:00
  </time>
  <span
    aria-hidden="true"
    class="absolute left-0 bottom-0 pointer-events-none h-[3px]"
    data-v-1cf33334=""
  />
</div>
`;
